File: quick-sort/QuickSort.js
var quickSort = (array) => {
    quickSortElement(array, 0, array.length - 1);
    return array;
}

var quickSortElement = (array, start, end) => {
    if(start >= end) { return ; }
    let left = start;
    let right = end;
    let pivot = array[Math.floor((start + end) / 2)];
    while (left <= right) {
        while(array[left] < pivot) { left++; }
        while(array[right] > pivot) { right--; }
        if(left <= right) {
            [array[left], array[right]] = [array[right], array[left]];
            left++; right--;
        }
    }
    quickSortElement(array, start, right);
    quickSortElement(array, left, end);
}

exports.quickSortElement = quickSortElement;
exports.quickSort = quickSort;

File: quick-sort/QuickSort.swift
class QuickSort {
  func quickSort(_ array: inout [Int]) -> [Int] {
    quickSort(&array, 0, array.count - 1)
    return array
  }

  func quickSort(_ array: inout[Int], _ start: Int, _ end: Int) {
    if(start >= end) { return }
    var left = start
    var right = end
    var pivot = array[(start + end) / 2]
    while (left <= right) {
      while(array[left] < pivot) { left += 1}
      while(array[right] > pivot) { right -= 1 }
      if(left <= right) {
        (array[left], array[right]) = (array[right], array[left])
        left += 1; right -= 1
      }
    }
    quickSort(&array, start, right)
    quickSort(&array, left, end)
  }
}